java/ql/src/experimental/Security/CWE/CWE-094/GroovyInjection.qhelp

```diff
@@ -0,0 +1,66 @@
+<!DOCTYPE qhelp PUBLIC "-//Semmle//qhelp//EN" "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>
+Apache Groovy is a powerful, optionally typed and dynamic language, 
+with static-typing and static compilation capabilities.
+
+It integrates smoothly with any Java program, 
+and immediately delivers to your application powerful features, 
+including scripting capabilities, Domain-Specific Language authoring, 
+runtime and compile-time meta-programming and functional programming.
+
+If a Groovy script is built using attacker-controlled data, 
+and then evaluated, then it may allow the attacker to achieve to Remote Code Execution.
+</p>
+</overview>
+
+<recommendation>
+<p>
+It is generally recommended to avoid using untrusted input in a Groovy evaluation.
+If it is not possible, use sandbox solution. Also you must take care about Groovy
+compile-time metaprogramming RCE case. It is possible to achieve to RCE by only compiling
+a Groovy script (see Groovy RCE Research on Jenkins by Orange Tsai).
+</p>
+</recommendation>
+
+<example>
+<p>
+The following example uses untrusted data to evaluate a Groovy script.
+</p>
+<sample src="GroovyInjectionBad.java" />
+
+<p>
+The following example uses sandbox protection before running Groovy script.
+Source code is used from https://melix.github.io/blog/2015/03/sandboxing.html.
```

Comment:
This code is licenses CC-BY-NC-SA, where NC is for Non-Commercial. While this is an open repository, CodeQL is a commercial product, so I am unsure if this is allowed. Could you replace their code by either (a) excluding the sandboxing .groovy file and simply explaining that some sandboxing tool exists, or (b) replacing the code from that site for code that conveys the same idea but which is of your own devising.

Reply:
The whole sandbox logic is in that .groovy file. Can I only write explanation in qhelp file without java code example?
Also there is no a silver-bullet solution for Groovy code execution protection, the reader must do additional research besides qhelp file explanation.

Reply:
If the code required is long and complicated perhaps point to it as a reference rather than copy-paste it into our qhelp.